File: platform/android/MapLibreAndroid/src/main/java/org/maplibre/android/offline/OfflineRegionError.kt
package org.maplibre.android.offline

import androidx.annotation.Keep
import androidx.annotation.StringDef

/**
 * An Offline Region error
 */
class OfflineRegionError // For JNI use only // Constructors
@Keep private constructor( // Getters
    @field:ErrorReason
    @get:ErrorReason
    val reason: String,
    /**
     * / * An error message from the request handler, e.g. a server message or a system message
     * / * informing the user about the reason for the failure.
     */
    val message: String
) {
    /**
     * Error code, as a string, self-explanatory.
     */
    @StringDef(*[REASON_SUCCESS, REASON_NOT_FOUND, REASON_SERVER, REASON_CONNECTION, REASON_OTHER])
    @kotlin.annotation.Retention(AnnotationRetention.SOURCE)
    annotation class ErrorReason

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }
        if (o == null || javaClass != o.javaClass) {
            return false
        }
        val that = o as OfflineRegionError
        return if (reason != that.reason) {
            false
        } else {
            message == that.message
        }
    }

    override fun hashCode(): Int {
        var result = reason.hashCode()
        result = 31 * result + message.hashCode()
        return result
    }

    override fun toString(): String {
        return ("OfflineRegionError{" + "reason='" + reason + '\'' + ", message='" + message + '\'' + '}')
    }

    companion object {
        const val REASON_SUCCESS = "REASON_SUCCESS"
        const val REASON_NOT_FOUND = "REASON_NOT_FOUND"
        const val REASON_SERVER = "REASON_SERVER"
        const val REASON_CONNECTION = "REASON_CONNECTION"
        const val REASON_OTHER = "REASON_OTHER"
    }
}

File: platform/android/MapLibreAndroid/src/cpp/offline/offline_region_error.cpp
#include "offline_region_error.hpp"

namespace mln {
namespace android {

jni::Local<jni::Object<OfflineRegionError>> OfflineRegionError::New(jni::JNIEnv& env, mln::Response::Error error) {
    // Handle the value of reason independently of the underlying int value
    std::string reason;
    switch (error.reason) {
        case mln::Response::Error::Reason::Success:
            reason = "REASON_SUCCESS";
            break;
        case mln::Response::Error::Reason::NotFound:
            reason = "REASON_NOT_FOUND";
            break;
        case mln::Response::Error::Reason::Server:
            reason = "REASON_SERVER";
            break;
        case mln::Response::Error::Reason::Connection:
            reason = "REASON_CONNECTION";
            break;
        case mln::Response::Error::Reason::RateLimit:
            reason = "REASON_RATE_LIMIT";
            break;
        case mln::Response::Error::Reason::Other:
            reason = "REASON_OTHER";
            break;
    }

    static auto& javaClass = jni::Class<OfflineRegionError>::Singleton(env);
    static auto constructor = javaClass.GetConstructor<jni::String, jni::String>(env);

    return javaClass.New(
        env, constructor, jni::Make<jni::String>(env, reason), jni::Make<jni::String>(env, error.message));
}

void OfflineRegionError::registerNative(jni::JNIEnv& env) {
    jni::Class<OfflineRegionError>::Singleton(env);
}

} // namespace android
} // namespace mln
